sbom-utility-scripts/scripts/merge-cachi2-sboms-script/merge_cachi2_sboms.py

```diff
@@ -148,6 +299,199 @@ def _merge_tools_metadata(syft_sbom: dict[Any, Any], cachi2_sbom: dict[Any, Any]
+
+
+def merge_components(syft_sbom: dict, cachi2_sbom: dict) -> dict:
+    """Merge Cachi2 components into the Syft SBOM while removing duplicates."""
+    is_duplicate_component = _get_syft_component_filter(cachi2_sbom["components"])
+    filtered_syft_components = [c for c in syft_sbom.get("components", []) if not is_duplicate_component(c)]
+    return filtered_syft_components + cachi2_sbom["components"]
+
+
+def merge_external_refs(refs1, refs2):
+    """Merge SPDX external references while removing duplicates."""
+    ref_tuples = []
+    unique_refs2 = []
+
+    for _ref in refs1:
+        ref = _ref.copy()
+        if ref["referenceType"].lower() == "purl":
+            parsed_purl = PackageURL.from_string(ref["referenceLocator"])
+            purl_dict = parsed_purl.to_dict()
+            purl_dict["qualifiers"] = {}
+            parsed_purl = PackageURL(**purl_dict)
+            ref["referenceLocator"] = parsed_purl.to_string()
+
+        ref_tuples.append(
+            (
+                ref["referenceCategory"].lower(),
+                ref["referenceType"].lower(),
+                ref["referenceLocator"].lower(),
+            )
+        )
+
+    for _ref in refs2:
+        ref = _ref.copy()
+        if ref["referenceType"].lower() == "purl":
+            parsed_purl = PackageURL.from_string(ref["referenceLocator"])
+            purl_dict = parsed_purl.to_dict()
+            purl_dict["qualifiers"] = {}
+            parsed_purl = PackageURL(**purl_dict)
+            ref["referenceLocator"] = parsed_purl.to_string()
+        if (
+            ref["referenceCategory"].lower(),
+            ref["referenceType"].lower(),
+            ref["referenceLocator"].lower(),
+        ) not in ref_tuples:
+            unique_refs2.append(ref)
+    return [ref for ref in refs1 + unique_refs2]
+
+
+def merge_annotations(annotations1, annotations2):
+    """Merge SPDX package annotations."""
+    annotation_tuples = []
+    for annotation in annotations1:
+        annotation_tuples.append(
+            (
+                annotation["annotator"],
+                annotation["comment"],
+                annotation["annotationDate"],
+                annotation["annotationType"],
+            )
+        )
+    for annotation in annotations2:
+        annotation_tuples.append(
+            (
+                annotation["annotator"],
+                annotation["comment"],
+                annotation["annotationDate"],
+                annotation["annotationType"],
+            )
+        )
+    annotations = set(annotation_tuples)
+    return [
+        {
+            "annotator": annotation[0],
+            "comment": annotation[1],
+            "annotationDate": annotation[2],
+            "annotationType": annotation[3],
+        }
+        for annotation in annotations
+    ]
+
+
+def merge_relationships(relationships1, relationships2, packages):
+    """Merge SPDX relationships."""
+
+    def map_relationships(relationships):
```

Comment:
The method of looking for the Document SPDXID and the root SPDXID is similarly overcomplicated and with similar deficiencies as https://github.com/konflux-ci/build-tasks-dockerfiles/pull/163#discussion_r1863588887